theano/tensor/nnet/bn.py

```diff
@@ -73,3 +79,387 @@ def batch_normalization(inputs, gamma, beta, mean, std,
         raise ValueError(
             'mode must be either "low_mem", "high_mem"')
     return rval
+
+
+def batch_normalization_train(inputs, gamma, beta, mode='per-activation',
+                              epsilon=1e-4):
+    """
+    Performs batch normalization of the given inputs, using the mean and
+    variance of the inputs.
+
+    Parameters
+    ----------
+    mode : {'per-activation', 'spatial'}
+        Whether to normalize per activation or share normalization factors
+        across spatial dimensions (i.e., all dimensions past the second).
```

Comment:
I'd prefer to have an `axes` argument here, to freely specify the axes to be normalized over. Not all cases can be replaced with the cuDNN implementation then, but we shouldn't restrict the interface because of that.